[multi-asic] Added Support  for multi-asic for telemetry/gnmi server

abdosi opened the pull request:
What I did:-
This PR handles the changes as mention in the document:
[MultiasicTelemetry.docx](https://github.com/Azure/sonic-telemetry/files/6386778/MultiasicTelemetry.docx)

Summary of Changes:-
- Enhanced `sonic_db_config` package for multi-asic to read different namespace redis configuration. Added unit-test for same.
- Enhanced V2R Lookup for multi-asic by understanding namespace port belongs to
- Enhanced gNMI Server to initiate redis connection with all namespaces
- Enhanced gNMI Server `Get` and `Subscribe` tests for multi-namespace. Added `test_utils` package providing utility API's needed by test-cases. Also added multi-namespace specific `json` files into testdata
- Enhance parsing of `Target` in gNMI Request to understand namespace if present.
- Format the modified files using `go fmt`. Please ignore whitespaces diff as part of review. 
- Added new test case to verify gNMI Get on non-counter DB (eg: STATE_DB) and fixed issue in existing GET of not verifying the return value.
- Fixed the gNMI Subscribe test issue where `IntervalTimeTicker` function pointer not re-assigned after test case is done.

How I Verify:

- Manual Verification of Single and Multi-asic platforms. 

1. Non V2R Dataset. Query per asic/namespace. Result from multi-asic platform
```
root@str2-xxxx-acs-3:/# gnmi_get -target_addr localhost:50051 -xpath SWITCH_CAPABILITY/switch -xpath_target STATE_DB/asic3 -insecure
== getRequest:
prefix: <
  target: "STATE_DB/asic3"
>
path: <
  elem: <
    name: "SWITCH_CAPABILITY"
  >
  elem: <
    name: "switch"
  >
>
encoding: JSON_IETF

== getResponse:
notification: <
  timestamp: 1619548691457577658
  prefix: <
    target: "STATE_DB/asic3"
  >
  update: <
    path: <
      elem: <
        name: "SWITCH_CAPABILITY"
      >
      elem: <
        name: "switch"
      >
    >
    val: <
      json_ietf_val: "{\"ACL_ACTIONS|EGRESS\":\"PACKET_ACTION\",\"ACL_ACTIONS|INGRESS\":\"PACKET_ACTION,MIRROR_INGRESS_ACTION\",\"ACL_ACTION|PACKET_ACTION\":\"DROP,FORWARD\",\"MAX_NEXTHOP_GROUP_COUNT\":\"256\",\"MIRROR\":\"true\",\"MIRRORV6\":\"true\"}"
    >
  >
>
```
2. V2R Dataset. Query Interface Statistics. Result from multi-asic platform
```
root@str2-xxxx-acs-3:/# gnmi_get -target_addr localhost:50051 -xpath COUNTERS/Ethernet220 -xpath_target COUNTERS_DB -insecure
== getRequest:
prefix: <
  target: "COUNTERS_DB"
>
path: <
  elem: <
    name: "COUNTERS"
  >
  elem: <
    name: "Ethernet220"
  >
>
encoding: JSON_IETF

== getResponse:
notification: <
  timestamp: 1619548775670855765
  prefix: <
    target: "COUNTERS_DB"
  >
  update: <
    path: <
      elem: <
        name: "COUNTERS"
      >
      elem: <
        name: "Ethernet220"
      >
    >
    val: <
      json_ietf_val: "{\"SAI_PORT_STAT_ETHER_IN_PKTS_128_TO_255_OCTETS\":\"354\",\"SAI_PORT_STAT_ETHER_RX_OVERSIZE_PKTS\":\"0\",\"SAI_PORT_STAT_ETHER_STATS_TX_NO_ERRORS\":\"1333\",\"SAI_PORT_STAT_ETHER_TX_OVERSIZE_PKTS\":\"0\",\"SAI_PORT_STAT_IF_IN_BROADCAST_PKTS\":\"1\",\"SAI_PORT_STAT_IF_IN_DISCARDS\":\"0\",\"SAI_PORT_STAT_IF_IN_ERRORS\":\"0\",\"SAI_PORT_STAT_IF_IN_MULTICAST_PKTS\":\"353\",\"SAI_PORT_STAT_IF_IN_NON_UCAST_PKTS\":\"354\",\"SAI_PORT_STAT_IF_IN_OCTETS\":\"122016\",\"SAI_PORT_STAT_IF_IN_UCAST_PKTS\":\"810\",\"SAI_PORT_STAT_IF_IN_UNKNOWN_PROTOS\":\"0\",\"SAI_PORT_STAT_IF_OUT_BROADCAST_PKTS\":\"1\",\"SAI_PORT_STAT_IF_OUT_DISCARDS\":\"0\",\"SAI_PORT_STAT_IF_OUT_ERRORS\":\"0\",\"SAI_PORT_STAT_IF_OUT_MULTICAST_PKTS\":\"361\",\"SAI_PORT_STAT_IF_OUT_NON_UCAST_PKTS\":\"362\",\"SAI_PORT_STAT_IF_OUT_OCTETS\":\"1025148\",\"SAI_PORT_STAT_IF_OUT_QLEN\":\"0\",\"SAI_PORT_STAT_IF_OUT_UCAST_PKTS\":\"971\",\"SAI_PORT_STAT_IN_DROPPED_PKTS\":\"0\",\"SAI_PORT_STAT_IP_IN_UCAST_PKTS\":\"5\",\"SAI_PORT_STAT_OUT_DROPPED_PKTS\":\"0\",\"SAI_PORT_STAT_PAUSE_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PAUSE_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_0_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_0_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_1_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_1_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_2_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_2_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_3_ON2OFF_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_3_ON2OFF_RX_PKTS_last\":\"0\",\"SAI_PORT_STAT_PFC_3_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_3_RX_PKTS_last\":\"0\",\"SAI_PORT_STAT_PFC_3_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_4_ON2OFF_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_4_ON2OFF_RX_PKTS_last\":\"0\",\"SAI_PORT_STAT_PFC_4_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_4_RX_PKTS_last\":\"0\",\"SAI_PORT_STAT_PFC_4_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_5_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_5_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_6_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_6_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_7_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_7_TX_PKTS\":\"0\"}"
    >
  >
>
```
3. All Unit-test passing . Result for  `GET` and `SUBSCRIBE` with new `Namespace` specific test-cases also.
```
root@abdosi-ubuntu-vm0:/src/sonic-telemetry/gnmi_server# /usr/local/go/bin/go test -mod=vendor  -v github.com/Azure/sonic-telemetry/gnmi_server -run TestGnmiGet -count=1
=== RUN   TestGnmiGet
=== RUN   TestGnmiGet/Test_non-existing_path_Target
E0427 18:41:51.150714    9141 request_binder.go:318] Error in creating the target object: rpc error: code = InvalidArgument desc = no match found in *ocbinds.Device, for path element:"MyCounters" elem:{name:"MyCounters"}
E0427 18:41:51.151419    9141 request_binder.go:176] Error in creating the target object : rpc error: code = InvalidArgument desc = no match found in *ocbinds.Device, for path element:"MyCounters" elem:{name:"MyCounters"}
=== RUN   TestGnmiGet/Test_empty_path_target
=== RUN   TestGnmiGet/Get_valid_but_non-existing_node
=== RUN   TestGnmiGet/Get_COUNTERS_PORT_NAME_MAP
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet68
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet68_SAI_PORT_STAT_PFC_7_RX_PKTS
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet68_Pfcwd
=== RUN   TestGnmiGet/get_COUNTERS_(use_vendor_alias):Ethernet68/1
=== RUN   TestGnmiGet/get_COUNTERS_(use_vendor_alias):Ethernet68/1_SAI_PORT_STAT_PFC_7_RX_PKTS
=== RUN   TestGnmiGet/get_COUNTERS_(use_vendor_alias):Ethernet68/1_Pfcwd
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet*
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet*_SAI_PORT_STAT_PFC_7_RX_PKTS
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet*_Pfcwd
=== RUN   TestGnmiGet/get_State_DB_Data_for_SWITCH_CAPABILITY_switch
=== RUN   TestGnmiGet/get_osversion/build
=== RUN   TestGnmiGet/get_osversion/build_file_load_error
E0427 18:41:51.200114    9141 non_db_client.go:308] Failed to read '/etc/sonic/sonic_version.yml', Cannot access '/etc/sonic/sonic_version.yml'
=== RUN   TestGnmiGet/get_osversion/build_file_parse_error
E0427 18:41:51.200685    9141 non_db_client.go:315] Failed to parse '/etc/sonic/sonic_version.yml', yaml: unmarshal errors:
  line 1: cannot unmarshal !!str `not a v...` into client.SonicVersionInfo
=== RUN   TestGnmiGet/get_osversion/build_different_value
--- PASS: TestGnmiGet (1.49s)
    --- PASS: TestGnmiGet/Test_non-existing_path_Target (0.02s)
  --- PASS: TestGnmiGet/Test_empty_path_target (0.00s)                                                                                                                                                             --- PASS: TestGnmiGet/Get_valid_but_non-existing_node (0.01s)
    --- PASS: TestGnmiGet/Get_COUNTERS_PORT_NAME_MAP (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet68 (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet68_SAI_PORT_STAT_PFC_7_RX_PKTS (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet68_Pfcwd (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS_(use_vendor_alias):Ethernet68/1 (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS_(use_vendor_alias):Ethernet68/1_SAI_PORT_STAT_PFC_7_RX_PKTS (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS_(use_vendor_alias):Ethernet68/1_Pfcwd (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet* (0.01s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet*_SAI_PORT_STAT_PFC_7_RX_PKTS (0.01s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet*_Pfcwd (0.01s)
    --- PASS: TestGnmiGet/get_State_DB_Data_for_SWITCH_CAPABILITY_switch (0.00s)
    --- PASS: TestGnmiGet/get_osversion/build (0.00s)
    --- PASS: TestGnmiGet/get_osversion/build_file_load_error (0.00s)
    --- PASS: TestGnmiGet/get_osversion/build_file_parse_error (0.00s)
    --- PASS: TestGnmiGet/get_osversion/build_different_value (0.00s)
=== RUN   TestGnmiGetMultiNs
=== RUN   TestGnmiGetMultiNs/Test_non-existing_path_Target
E0427 18:41:56.007094    9141 request_binder.go:318] Error in creating the target object: rpc error: code = InvalidArgument desc = no match found in *ocbinds.Device, for path element:"MyCounters" elem:{name:"$yCounters"}
E0427 18:41:56.007123    9141 request_binder.go:176] Error in creating the target object : rpc error: code = InvalidArgument desc = no match found in *ocbinds.Device, for path element:"MyCounters" elem:{name:$MyCounters"}
=== RUN   TestGnmiGetMultiNs/Test_empty_path_target
=== RUN   TestGnmiGetMultiNs/Get_valid_but_non-existing_node
=== RUN   TestGnmiGetMultiNs/Get_COUNTERS_PORT_NAME_MAP
=== RUN   TestGnmiGetMultiNs/get_COUNTERS:Ethernet68
=== RUN   TestGnmiGetMultiNs/get_COUNTERS:Ethernet68_SAI_PORT_STAT_PFC_7_RX_PKTS
=== RUN   TestGnmiGetMultiNs/get_COUNTERS:Ethernet68_Pfcwd
=== RUN   TestGnmiGetMultiNs/get_COUNTERS_(use_vendor_alias):Ethernet68/1
=== RUN   TestGnmiGetMultiNs/get_COUNTERS_(use_vendor_alias):Ethernet68/1_SAI_PORT_STAT_PFC_7_RX_PKTS
=== RUN   TestGnmiGetMultiNs/get_COUNTERS_(use_vendor_alias):Ethernet68/1_Pfcwd
=== RUN   TestGnmiGetMultiNs/get_COUNTERS:Ethernet*
=== RUN   TestGnmiGetMultiNs/get_COUNTERS:Ethernet*_SAI_PORT_STAT_PFC_7_RX_PKTS
=== RUN   TestGnmiGetMultiNs/get_COUNTERS:Ethernet*_Pfcwd
=== RUN   TestGnmiGetMultiNs/get_State_DB_Data_for_SWITCH_CAPABILITY_switch
=== RUN   TestGnmiGetMultiNs/get_osversion/build
=== RUN   TestGnmiGetMultiNs/get_osversion/build_file_load_error
E0427 18:41:56.041761    9141 non_db_client.go:308] Failed to read '/etc/sonic/sonic_version.yml', Cannot access '/etc/sonic/sonic_version.yml'
=== RUN   TestGnmiGetMultiNs/get_osversion/build_file_parse_error
E0427 18:41:56.042256    9141 non_db_client.go:315] Failed to parse '/etc/sonic/sonic_version.yml', yaml: unmarshal errors:
  line 1: cannot unmarshal !!str `not a v...` into client.SonicVersionInfo
=== RUN   TestGnmiGetMultiNs/get_osversion/build_different_value
--- PASS: TestGnmiGetMultiNs (4.84s)
    --- PASS: TestGnmiGetMultiNs/Test_non-existing_path_Target (0.01s)
    --- PASS: TestGnmiGetMultiNs/Test_empty_path_target (0.00s)
    --- PASS: TestGnmiGetMultiNs/Get_valid_but_non-existing_node (0.00s)
    --- PASS: TestGnmiGetMultiNs/Get_COUNTERS_PORT_NAME_MAP (0.00s)
    --- PASS: TestGnmiGetMultiNs/get_COUNTERS:Ethernet68 (0.00s)
    --- PASS: TestGnmiGetMultiNs/get_COUNTERS:Ethernet68_SAI_PORT_STAT_PFC_7_RX_PKTS (0.00s)
    --- PASS: TestGnmiGetMultiNs/get_COUNTERS:Ethernet68_Pfcwd (0.00s)
    --- PASS: TestGnmiGetMultiNs/get_COUNTERS_(use_vendor_alias):Ethernet68/1 (0.00s)
    --- PASS: TestGnmiGetMultiNs/get_COUNTERS_(use_vendor_alias):Ethernet68/1_SAI_PORT_STAT_PFC_7_RX_PKTS (0.00s)
    --- PASS: TestGnmiGetMultiNs/get_COUNTERS_(use_vendor_alias):Ethernet68/1_Pfcwd (0.00s)
    --- PASS: TestGnmiGetMultiNs/get_COUNTERS:Ethernet* (0.01s)
    --- PASS: TestGnmiGetMultiNs/get_COUNTERS:Ethernet*_SAI_PORT_STAT_PFC_7_RX_PKTS (0.00s)
    --- PASS: TestGnmiGetMultiNs/get_COUNTERS:Ethernet*_Pfcwd (0.01s)
    --- PASS: TestGnmiGetMultiNs/get_State_DB_Data_for_SWITCH_CAPABILITY_switch (0.00s)
    --- PASS: TestGnmiGetMultiNs/get_osversion/build (0.00s)
    --- PASS: TestGnmiGetMultiNs/get_osversion/build_file_load_error (0.00s)
    --- PASS: TestGnmiGetMultiNs/get_osversion/build_file_parse_error (0.00s)
    --- PASS: TestGnmiGetMultiNs/get_osversion/build_different_value (0.00s)
=== RUN   TestGnmiGetTranslib
=== RUN   TestGnmiGetTranslib/Get_OC_Interfaces
=== RUN   TestGnmiGetTranslib/Get_OC_Interface
=== RUN   TestGnmiGetTranslib/Get_OC_Interface_admin-status
=== RUN   TestGnmiGetTranslib/Get_OC_Interface_ifindex
=== RUN   TestGnmiGetTranslib/Get_OC_Interface_mtu
--- PASS: TestGnmiGetTranslib (1.81s)
    --- PASS: TestGnmiGetTranslib/Get_OC_Interfaces (0.04s)
    --- PASS: TestGnmiGetTranslib/Get_OC_Interface (0.00s)
    --- PASS: TestGnmiGetTranslib/Get_OC_Interface_admin-status (0.00s)
    --- PASS: TestGnmiGetTranslib/Get_OC_Interface_ifindex (0.00s)
    --- PASS: TestGnmiGetTranslib/Get_OC_Interface_mtu (0.00s)
PASS
ok      github.com/Azure/sonic-telemetry/gnmi_server    14.346s      
```
```
root@abdosi-ubuntu-vm0:/src/sonic-telemetry/gnmi_server# /usr/local/go/bin/go test -mod=vendor  -v github.com/Azure/sonic-telemetry/gnmi_server -run TestGnmiSubscribe -count=1
    --- PASS: TestGnmiSubscribe/stream_query_for_table_COUNTERS_PORT_NAME_MAP_with_new_test_field_field (2.00s)
    --- PASS: TestGnmiSubscribe/stream_query_for_table_key_Ethernet68_with_new_test_field_field (3.00s)
    --- PASS: TestGnmiSubscribe/(use_vendor_alias)_stream_query_for_table_key_Ethernet68/1_with_new_test_field_field (3.00s)
    --- PASS: TestGnmiSubscribe/stream_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_update_of_field_value (3.00s)
    --- PASS: TestGnmiSubscribe/(use_vendor_alias)_stream_query_for_COUNTERS/[Ethernet68/1]/SAI_PORT_STAT_PFC_7_RX_PKTS_with_update_of_field_value (3.00s)
    --- PASS: TestGnmiSubscribe/stream_query_for_COUNTERS/Ethernet68/Pfcwd_with_update_of_field_value (3.00s)
    --- PASS: TestGnmiSubscribe/(use_vendor_alias)_stream_query_for_COUNTERS/[Ethernet68/1]/Pfcwd_with_update_of_field_value (3.00s)
    --- PASS: TestGnmiSubscribe/stream_query_for_table_key_Ethernet*_with_new_test_field_field_on_Ethernet68 (3.00s)
    --- PASS: TestGnmiSubscribe/stream_query_for_table_key_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_update (2.00s)
    --- PASS: TestGnmiSubscribe/stream_query_for_table_key_Ethernet*/Pfcwd_with_field_value_update (2.00s)
    --- PASS: TestGnmiSubscribe/poll_query_for_table_COUNTERS_PORT_NAME_MAP_with_new_field_test_field (2.01s)
    --- PASS: TestGnmiSubscribe/poll_query_for_table_COUNTERS_PORT_NAME_MAP_with_test_field_delete (2.00s)
    --- PASS: TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_change (2.00s)
    --- PASS: TestGnmiSubscribe/(use_vendor_alias)_poll_query_for_COUNTERS/[Ethernet68/1]/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_change (2.00s)
    --- PASS: TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet68/Pfcwd_with_field_value_change (2.00s)
    --- PASS: TestGnmiSubscribe/(use_vendor_alias)_poll_query_for_COUNTERS/[Ethernet68/1]/Pfcwd_with_field_value_change (2.00s)
    --- PASS: TestGnmiSubscribe/poll_query_for_table_key_Ethernet*_with_Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_field_value_change (2.02s)
    --- PASS: TestGnmiSubscribe/poll_query_for_table_key_field_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_field_value_change (2.01s)
    --- PASS: TestGnmiSubscribe/poll_query_for_table_key_field_Etherenet*/Pfcwd_with_Ethernet68:3/PFC_WD_QUEUE_STATS_DEADLOCK_DETECTED_field_value_change (2.03s)
    --- PASS: TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet*/Queues (1.10s)
    --- PASS: TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet68/Queues_with_field_value_change (2.01s)
    --- PASS: TestGnmiSubscribe/(use_vendor_alias)_poll_query_for_COUNTERS/Ethernet68/Queues_with_field_value_change (2.01s)
    --- PASS: TestGnmiSubscribe/use_invalid_sample_interval (1.00s)
    --- PASS: TestGnmiSubscribe/sample_stream_query_for_table_key_Ethernet68_with_new_test_field_field (2.00s)
    --- PASS: TestGnmiSubscribe/sample_stream_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_2_updates (3.00s)
    --- PASS: TestGnmiSubscribe/(use_vendor_alias)_sample_stream_query_for_table_key_Ethernet68/1_with_new_test_field_field (3.00s)
    --- PASS: TestGnmiSubscribe/sample_stream_query_for_COUNTERS/Ethernet68/Pfcwd_with_update_of_field_value (3.00s)
    --- PASS: TestGnmiSubscribe/(use_vendor_alias)_sample_stream_query_for_COUNTERS/[Ethernet68/1]/Pfcwd_with_update_of_field_value (3.00s)
    --- PASS: TestGnmiSubscribe/sample_stream_query_for_table_key_Ethernet*_with_new_test_field_field_on_Ethernet68 (2.00s)
    --- PASS: TestGnmiSubscribe/(updates_only)_sample_stream_query_for_table_key_Ethernet*_with_new_test_field_field_on_Ethernet68 (4.00s)
    --- PASS: TestGnmiSubscribe/sample_stream_query_for_table_key_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_update (2.00s)
    --- PASS: TestGnmiSubscribe/sample_stream_query_for_table_key_Ethernet*/Pfcwd_with_field_value_update (2.00s)
    --- PASS: TestGnmiSubscribe/(update_only)_sample_stream_query_for_table_key_Ethernet*/Pfcwd_with_field_value_update (4.00s)
--- PASS: TestGnmiSubscribeMultiNs (115.26s)
    --- PASS: TestGnmiSubscribeMultiNs/stream_query_for_table_COUNTERS_PORT_NAME_MAP_with_new_test_field_field (2.00s)
    --- PASS: TestGnmiSubscribeMultiNs/stream_query_for_table_key_Ethernet68_with_new_test_field_field (3.00s)
    --- PASS: TestGnmiSubscribeMultiNs/(use_vendor_alias)_stream_query_for_table_key_Ethernet68/1_with_new_test_field_field (3.00s)
    --- PASS: TestGnmiSubscribeMultiNs/stream_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_update_of_field_value (3.00s)
    --- PASS: TestGnmiSubscribeMultiNs/(use_vendor_alias)_stream_query_for_COUNTERS/[Ethernet68/1]/SAI_PORT_STAT_PFC_7_RX_PKTS_with_update_of_field_value (3.00s)
    --- PASS: TestGnmiSubscribeMultiNs/stream_query_for_COUNTERS/Ethernet68/Pfcwd_with_update_of_field_value (3.00s)
    --- PASS: TestGnmiSubscribeMultiNs/(use_vendor_alias)_stream_query_for_COUNTERS/[Ethernet68/1]/Pfcwd_with_update_of_field_value (3.00s)
    --- PASS: TestGnmiSubscribeMultiNs/stream_query_for_table_key_Ethernet*_with_new_test_field_field_on_Ethernet68 (3.00s)
    --- PASS: TestGnmiSubscribeMultiNs/stream_query_for_table_key_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_update (2.00s)
    --- PASS: TestGnmiSubscribeMultiNs/stream_query_for_table_key_Ethernet*/Pfcwd_with_field_value_update (2.00s)
    --- PASS: TestGnmiSubscribeMultiNs/poll_query_for_table_COUNTERS_PORT_NAME_MAP_with_new_field_test_field (2.00s)
    --- PASS: TestGnmiSubscribeMultiNs/poll_query_for_table_COUNTERS_PORT_NAME_MAP_with_test_field_delete (2.00s)
    --- PASS: TestGnmiSubscribeMultiNs/poll_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_change (2.00s)
    --- PASS: TestGnmiSubscribeMultiNs/(use_vendor_alias)_poll_query_for_COUNTERS/[Ethernet68/1]/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_change (2.00s)
    --- PASS: TestGnmiSubscribeMultiNs/poll_query_for_COUNTERS/Ethernet68/Pfcwd_with_field_value_change (2.00s)
    --- PASS: TestGnmiSubscribeMultiNs/(use_vendor_alias)_poll_query_for_COUNTERS/[Ethernet68/1]/Pfcwd_with_field_value_change (2.00s)
    --- PASS: TestGnmiSubscribeMultiNs/poll_query_for_table_key_Ethernet*_with_Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_field_value_change (2.02s)
    --- PASS: TestGnmiSubscribeMultiNs/poll_query_for_table_key_field_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_field_value_change (2.01s)
    --- PASS: TestGnmiSubscribeMultiNs/poll_query_for_table_key_field_Etherenet*/Pfcwd_with_Ethernet68:3/PFC_WD_QUEUE_STATS_DEADLOCK_DETECTED_field_value_change (2.03s)
    --- PASS: TestGnmiSubscribeMultiNs/poll_query_for_COUNTERS/Ethernet*/Queues (1.08s)
    --- PASS: TestGnmiSubscribeMultiNs/poll_query_for_COUNTERS/Ethernet68/Queues_with_field_value_change (2.01s)
    --- PASS: TestGnmiSubscribeMultiNs/(use_vendor_alias)_poll_query_for_COUNTERS/Ethernet68/Queues_with_field_value_change (2.01s)
    --- PASS: TestGnmiSubscribeMultiNs/use_invalid_sample_interval (1.00s)
    --- PASS: TestGnmiSubscribeMultiNs/sample_stream_query_for_table_key_Ethernet68_with_new_test_field_field (2.00s)
    --- PASS: TestGnmiSubscribeMultiNs/sample_stream_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_2_updates (3.00s)
    --- PASS: TestGnmiSubscribeMultiNs/(use_vendor_alias)_sample_stream_query_for_table_key_Ethernet68/1_with_new_test_field_field (3.00s)
    --- PASS: TestGnmiSubscribeMultiNs/sample_stream_query_for_COUNTERS/Ethernet68/Pfcwd_with_update_of_field_value (3.00s)
    --- PASS: TestGnmiSubscribeMultiNs/(use_vendor_alias)_sample_stream_query_for_COUNTERS/[Ethernet68/1]/Pfcwd_with_update_of_field_value (3.00s)
    --- PASS: TestGnmiSubscribeMultiNs/sample_stream_query_for_table_key_Ethernet*_with_new_test_field_field_on_Ethernet68 (2.00s)
    --- PASS: TestGnmiSubscribeMultiNs/(updates_only)_sample_stream_query_for_table_key_Ethernet*_with_new_test_field_field_on_Ethernet68 (4.00s)
    --- PASS: TestGnmiSubscribeMultiNs/sample_stream_query_for_table_key_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_update (2.00s)
    --- PASS: TestGnmiSubscribeMultiNs/sample_stream_query_for_table_key_Ethernet*/Pfcwd_with_field_value_update (2.00s)
    --- PASS: TestGnmiSubscribeMultiNs/(update_only)_sample_stream_query_for_table_key_Ethernet*/Pfcwd_with_field_value_update (4.00s)
PASS
ok      github.com/Azure/sonic-telemetry/gnmi_server    230.306s
```
4. Newly added `sonic_db_config` testcases passed.
```
root@abdosi-ubuntu-vm0:/src/sonic-telemetry/gnmi_server# /usr/local/go/bin/go test -mod=vendor  -v github.com/Azure/sonic-telemetry/sonic_db_config -count=1
=== RUN   TestGetDb
=== RUN   TestGetDb/Id
=== RUN   TestGetDb/Sock
=== RUN   TestGetDb/AllNamespaces
=== RUN   TestGetDb/TcpAddr
--- PASS: TestGetDb (0.00s)
    --- PASS: TestGetDb/Id (0.00s)
    --- PASS: TestGetDb/Sock (0.00s)
    --- PASS: TestGetDb/AllNamespaces (0.00s)
    --- PASS: TestGetDb/TcpAddr (0.00s)
=== RUN   TestGetDbMultiNs
=== RUN   TestGetDbMultiNs/Id
=== RUN   TestGetDbMultiNs/Sock
=== RUN   TestGetDbMultiNs/AllNamespaces
=== RUN   TestGetDbMultiNs/TcpAddr
--- PASS: TestGetDbMultiNs (0.00s)
    --- PASS: TestGetDbMultiNs/Id (0.00s)
    --- PASS: TestGetDbMultiNs/Sock (0.00s)
    --- PASS: TestGetDbMultiNs/AllNamespaces (0.00s)
    --- PASS: TestGetDbMultiNs/TcpAddr (0.00s)
PASS
ok      github.com/Azure/sonic-telemetry/sonic_db_config        0.005s
```
5. Verified no impact on dial-out client (Support Single namespace/asic only)
```
root@abdosi-ubuntu-vm0:/src/sonic-telemetry/gnmi_server# /usr/local/go/bin/go test -mod=vendor  -v github.com/Azure/sonic-telemetry/dialout/dialout_client -count=1
=== RUN   TestGNMIDialOutPublish
=== RUN   TestGNMIDialOutPublish/DialOut_to_first_collector_in_stream_mode_and_synced
=== RUN   TestGNMIDialOutPublish/DialOut_to_second_collector_in_stream_mode_upon_failure_of_first_collector
--- PASS: TestGNMIDialOutPublish (11.38s)
    --- PASS: TestGNMIDialOutPublish/DialOut_to_first_collector_in_stream_mode_and_synced (0.51s)
    --- PASS: TestGNMIDialOutPublish/DialOut_to_second_collector_in_stream_mode_upon_failure_of_first_collector (7.51s)
PASS
ok      github.com/Azure/sonic-telemetry/dialout/dialout_client 17.584s

```

abdosi commented:
@macikgozwa Somehow cannot add you in reviewers so tagging you.

qiluo-msft commented on sonic_data_client/db_client.go:
```diff
@@ -367,52 +368,92 @@ func ValToResp(val Value) (*gnmipb.SubscribeResponse, error) {
+	}
+	return redis_client_map
+}
+func IsTargetDb(target string) (string, bool, string, bool) {
```
>IsTargetDb [](http://example.com/codeflow?start=5&length=10)

Could you explain the purpose of this function and each paramter and return value?

abdosi replied:
This function get target present in GNMI Request and
returns: 1. DbName (string) 2. Is DbName valid (bool)
             3. DbNamespace (string) 4. Is DbNamespace present in Target

qiluo-msft replied:
What is "target present"?

abdosi replied:
Hi @qiluo-msft For eg: In below gNMI request value associated with `-xpath_target`  is target which is `STATE_DB/asic3`.
So above function will return (`state_db`, `true`, `asic3`, `true`)


root@str2-xxxx-acs-3:/# gnmi_get -target_addr localhost:50051 -xpath SWITCH_CAPABILITY/switch -xpath_target STATE_DB/asic3 -insecure
== getRequest:

qiluo-msft commented on Makefile:
```diff
@@ -60,12 +60,14 @@ check:
 clean:
 	$(RM) -r build
 	$(RM) -r vendor
+	$(RM) -r ${DBDIR}
```
>DBDIR [](http://example.com/codeflow?start=12&length=5)

this is created by `sudo`, could you clean without `sudo`?

abdosi replied:
yes updated and needed in Azure Pipeline setup. In local docker it was fine as running as `root`  so got missed.

qiluo-msft commented on sonic_data_client/db_client.go:
```diff
@@ -367,52 +368,92 @@ func ValToResp(val Value) (*gnmipb.SubscribeResponse, error) {
+	var separator string = sdcfg.GetDbSeparator(target, ns)
 	return separator, nil
 }
+func GetRedisClientsForDb(target string) map[string]*redis.Client {
```
>func [](http://example.com/codeflow?start=0&length=4)

Add one blank line between functions.

abdosi replied:
done

qiluo-msft commented:
@pra-moh @hui-ma Could you help review?

qiluo-msft reviewed (approved):
LGTM

abdosi commented on sonic_data_client/db_client.go:
```diff
@@ -447,19 +493,25 @@ func populateDbtablePath(prefix, path *gnmipb.Path, pathG2S *map[*gnmipb.Path][]
-		return fmt.Errorf("Invalid target name %v", target)
+		return fmt.Errorf("Invalid target dbNameSpace %v", targetDbNameSpace)
+	}
+	// Verify it is a valid db name
```
either way should be fine. However flip the condition as it save one function call.

abdosi commented on sonic_data_client/db_client.go:
```diff
@@ -493,6 +548,11 @@ func populateDbtablePath(prefix, path *gnmipb.Path, pathG2S *map[*gnmipb.Path][]
 
+	redisDb, ok := Target2RedisDb[tblPath.dbNamespace][tblPath.dbName]
+	if !ok {
+		return fmt.Errorf("Redis Client not present for dbName %v dbNamespace", targetDbName, dbNamespace)
```
thanks.. updated.

abdosi commented on sonic_data_client/virtual_db.go:
```diff
@@ -122,145 +123,161 @@ func getPfcwdMap() (map[string]map[string]string, error) {
+		separator, _ := GetTableKeySeparator(dbName, namespace)
+		_, err := redisDb.Ping().Result()
+		if err != nil {
+			log.V(1).Infof("Can not connect to %v, err: %v", dbName, err)
```
Updated all log/error messages to print namespace also

abdosi commented on sonic_db_config/db_config.go:
```diff
@@ -1,135 +1,237 @@
+	}
+	db_list, ok := sonic_db_config[ns]["DATABASES"].(map[string]interface{})
+	if !ok {
+		panic(fmt.Errorf("DATABASES' is not valid key in database_config.json file for namespace %q!", ns))
```
updated to use `%v`

abdosi commented on sonic_db_config/db_config.go:
```diff
@@ -1,135 +1,237 @@
+		if len(ns_to_cfgfile_map) > 1 {
+			sonic_db_multi_namespace = true
+		} else {
+			sonic_db_multi_namespace = false
```
yes in that case it will be same as regular case where there is only one redis instance running in host. It is not an error case.
These files are static are built into sonic image based on platform type so can't/shouldn't be change run-time.

abdosi commented on sonic_db_config/db_config.go:
```diff
@@ -1,135 +1,237 @@
+	} else if errors.Is(err, os.ErrNotExist) {
+		// Ref: https://stackoverflow.com/questions/23452157/how-do-i-check-for-specific-types-of-error-among-those-returned-by-ioutil-readfi
+		ns_to_cfgfile_map[SONIC_DEFAULT_NAMESPACE] = SONIC_DB_CONFIG_FILE
+		sonic_db_multi_namespace = false
```
for regular case/existing behavior this information comes from 
`SONIC_DB_CONFIG_FILE        string = "/var/run/redis/sonic-db/database_config.json"`

abdosi commented on sonic_data_client/db_client.go:
```diff
@@ -367,52 +368,97 @@ func ValToResp(val Value) (*gnmipb.SubscribeResponse, error) {
+	dbNameSpaceExist := false
+	dbNamespace := sdcfg.GetDbDefaultNamespace()
+
+	if len(targetname) > 2 {
```
Added the log message. Subsequent user of this function return appropriate error back to caller.

anshuv-mfst commented:
Hi @reshmaintel - please assign reviewers from Intel team, thanks.

hui-ma commented:
"If the “/ASICX” is not present, then request will map to default namespace (existing behavior will be achieved", does this rule also 
apply to counter_DB? or the counter_DB has its own rule for namespace?

reshmaintel commented:
Hi @akokhan could you please review this PR. Thanks

abdosi commented:
> "If the “/ASICX” is not present, then request will map to default namespace (existing behavior will be achieved", does this rule also
> apply to counter_DB? or the counter_DB has its own rule for namespace? From the example in your test result, don't know how to identify the interface from different ASIC.

@hui-ma for the counter DB we support V2R mapping dataset for which this PR (as mention in design document) get information from all ASIC's counter DB.  As of now this PR don't support passing /ASICx for Counter Db V2R set and log error for this case. However in future this is extendable when Counter DB has any dataset that needs it.

abdosi commented:
@macikgozwa can you please review/approve again.

abdosi commented:
@hui-ma and @@macikgozwa  is it ok to approve/merge this . Can you please check.

novikauanton commented on sonic_db_config/db_config.go:
```diff
@@ -1,135 +1,237 @@
+	sonic_db_init = true
+}
+
+func Init() {
```
why do the package init function replaced with the custom Init? 
And this function doesn't do actual initialization, it just set var to false, why do we need to call this function, probably it's better to initialize `sonic_db_init ` during declaration.
`var sonic_db_init  bool = false`

abdosi replied:
This was needed to initialize DB again as part of unit-test so that we can start testing multi-asic once single asic is done.
For example in function `TestGnmiGetMultiNs()` we call `Init()`. Since this make sonic_db_init  as false we will re-initialize the DB again.

Use of init() was existing design so not changing that.

novikauanton replied:
So this function doesn't do actual initialization, it does `reset` instead, and so probably it's better to give it appropriate name.
We do not have any reasons to leave `sonic_db_init ` uninitialized, and it can be initialized during declaration.

novikauanton commented on sonic_db_config/db_config.go:
```diff
@@ -1,135 +1,237 @@
+	}
+	ns_to_cfgfile_map := make(map[string]string)
+	// Ref: https://stackoverflow.com/questions/14928826/passing-pointers-to-maps-in-golang
+	DbGetNamespaceAndConfigFile(ns_to_cfgfile_map)
```
Seems that logic inside this function can be executed during the auto package initialization with `init`, and it can simplify code and remove 
```
if !sonic_db_init {
    DbInit()	
}
```
from each function call. 
Do we really need to skip possible errors with config files and handle them at runtime?
Or probably it's better to explicitly show errors and panic on library loading and avoid run lib with invalid config.

abdosi replied:
existing design. not changing as part of this PR.

novikauanton commented on sonic_db_config/db_config.go:
```diff
@@ -1,135 +1,237 @@
+		}
+		ns_list = append(ns_list, ns)
+	}
+	return ns_list, len(ns_list)
```
1) `len` is golang builtin function, that could be called on array/string without overhead and it's a well known way to get elements count with `len`
2) the result is not used after calls of this function
So as it is a builtin and it is ignored it is redundant

abdosi replied:
updated not to return length as it is not used.

novikauanton commented on sonic_db_config/db_config.go:
```diff
@@ -1,135 +1,237 @@
+	if !sonic_db_init {
+		DbInit()
+	}
+	db_list, ok := sonic_db_config[ns]["DATABASES"].(map[string]interface{})
```
Is it a possible case that target ns is not present in the map?
```
db_list, ok := sonic_db_config[ns]
ok == false
```

abdosi replied:
that is not possible. we read name space from database_config.json file so if that does not contain ns we will hit some other error before.

novikauanton commented on sonic_db_config/db_config.go:
```diff
@@ -1,135 +1,237 @@
+	return hostname + ":" + strconv.Itoa(port)
+}
+
+func DbGetNamespaceAndConfigFile(ns_to_cfgfile_map map[string]string) {
```
some functions use called `PrefixDb` notation and another part using `DbSuffix` notation.
For example `GetDbTcpAddr` and `DbGetNamespaceAndConfigFile`.
What is the accepted naming conversion in the project?

abdosi replied:
changes are to make sure backward compatibility so  existing fuction do not break.
No particuar design here.

novikauanton commented on sonic_db_config/db_config.go:
```diff
@@ -1,135 +1,237 @@
+		if err != nil {
+			panic(err)
+		}
+		var sonic_db_global_config = make(map[string]interface{})
```
`:=` ?

abdosi replied:
updated

novikauanton commented on sonic_db_config/db_config.go:
```diff
@@ -1,135 +1,237 @@
+		if err != nil {
+			panic(err)
+		}
+		var db_include_file string
```
This var is not used outside the cycle and doesn't store any state, why does this var declared outside the for cycle?
what is the purpose of this var, it is used the next line after its initialization?

abdosi replied:
updated.

novikauanton commented on sonic_db_config/db_config_test.go:
```diff
@@ -0,0 +1,78 @@
+	t.Run("Sock", func(t *testing.T) {
+		sock_path := GetDbSock("CONFIG_DB", GetDbDefaultNamespace())
+		if sock_path != "/var/run/redis/redis.sock" {
+			t.Fatalf(`Sock("") = %q, want "", error`, sock_path)
```
want "/var/run/redis/redis.sock" ?

abdosi replied:
thanks updated.

novikauanton commented on sonic_db_config/db_config_test.go:
```diff
@@ -0,0 +1,78 @@
+		if len(ns_list) != 1 {
+			t.Fatalf(`AllNamespaces("") = %q, want "1", error`, len(ns_list))
+		}
+		if !(ns_list[0] == GetDbDefaultNamespace()) {
```
`!(ns_list[0] == GetDbDefaultNamespace())` this can be simplified with operator `!=`

abdosi replied:
updated

novikauanton commented on sonic_db_config/db_config_test.go:
```diff
@@ -0,0 +1,78 @@
+	t.Run("Sock", func(t *testing.T) {
+		sock_path := GetDbSock("CONFIG_DB", "asic0")
+		if sock_path != "/var/run/redis0/redis.sock" {
+			t.Fatalf(`Sock("") = %q, want "", error`, sock_path)
```
`want "/var/run/redis0/redis.sock"`

abdosi replied:
thanks updated

novikauanton commented on sonic_db_config/db_config_test.go:
```diff
@@ -0,0 +1,78 @@
+		if len(ns_list) != 2 {
+			t.Fatalf(`AllNamespaces("") = %q, want "2", error`, len(ns_list))
+		}
+		if !((ns_list[0] == GetDbDefaultNamespace() && ns_list[1] == "asic0") || (ns_list[0] == "asic0" && ns_list[1] == GetDbDefaultNamespace())) {
```
https://pkg.go.dev/github.com/stretchr/testify/assert#ElementsMatch

abdosi replied:
thanks as of now only 2 condition so keep it like this. will explore this later

novikauanton commented on sonic_data_client/db_client.go:
```diff
@@ -367,52 +368,98 @@ func ValToResp(val Value) (*gnmipb.SubscribeResponse, error) {
-					DB:          int(dbn),
-					DialTimeout: 0,
-				})
+	if UseRedisLocalTcpPort {
```
1) Code nesting could be reduced.
```
if !UseRedisLocalTcpPort {
    return
}
```
2) Should `else` case be handled to?

abdosi replied:
updated for 1st point. For 2nd point  this function used only from perspective for unit-test. `else` is not needed if we somehow reach that condition  unit-test will fail.

novikauanton commented on sonic_data_client/db_client.go:
```diff
@@ -367,52 +368,98 @@ func ValToResp(val Value) (*gnmipb.SubscribeResponse, error) {
+			for dbName, dbn := range spb.Target_value {
+				if dbName != "OTHERS" {
+					// DB connector for direct redis operation
+					var redisDb *redis.Client
```
what is the purpose of this var?
If it is present to reduce code nesting, code nesting can be reduced in case of adding next statement as the first cycle statement
```
if dbName == "OTHERS" {
    continue
}
```

abdosi replied:
removed `var` and using `:=`

abdosi commented:
@hui-ma can we review/approve this ?

novikauanton commented on sonic_data_client/virtual_db.go:
```diff
@@ -419,7 +468,7 @@ func v2rEthPortPfcwdStats(paths []string) ([]tablePath, error) {
 // [COUNTER_DB COUNTERS Ethernet* Queues] or [COUNTER_DB COUNTERS Ethernet68 Queues]
 func v2rEthPortQueStats(paths []string) ([]tablePath, error) {
-	separator, _ := GetTableKeySeparator(paths[DbIdx])
+	separator, _ := GetTableKeySeparator(paths[DbIdx], "")
```
if the blank string is intended to get separator for default namespace, it's better to use constant for default database `sonic_db_config.db_config.SONIC_DEFAULT_NAMESPACE` or `GetDbDefaultNamespace()` from the same package

abdosi replied:
even though value is same the SONIC_DEFAULT_NAMESPACE  is used for some other purpose and has different meaning. To replace constant we will need to use other macro. This can be enhanced as part of other pr regarding cleanup the code

macikgozwa commented:
This is a comment for `db_client.go` line 132:
```
	if prefix.GetTarget() == "COUNTERS_DB" {
		err = initCountersPortNameMap()
		if err != nil {
			return nil, err
		}
               
               // Code to initialize other maps...
	}

```

This code wouldn't run in case target has a asic name segment. I recall `COUNTERS_DB` is not supported for multi asic queries.
However, `COUNTERS_DB/asic0` doesn't give me an error:

```
update: <
  timestamp: 1623266624421030390
  prefix: <
    target: "COUNTERS_DB/asic0"
  >
  update: <
    path: <
      elem: <
        name: "COUNTERS"
      >
      elem: <
        name: "Ethernet0"
      >
    >
    val: <
      string_val: ""
    >
  >
>

sync_response: true
```

Compare to the query without asic suffix (on the same environment):
```
update: <
  timestamp: 1623266666110051137
  prefix: <
    target: "COUNTERS_DB"
  >
  update: <
    path: <
      elem: <
        name: "COUNTERS"
      >
      elem: <
        name: "Ethernet0"
      >
    >
    val: <
      json_ietf_val: "{\"SAI_PORT_STAT_ETHER_IN_PKTS_128_TO_255_OCTETS\":\"0\",\"SAI_PORT_STAT_ETHER_RX_OVERSIZE_PKTS\":\"0\",\"SAI_PORT_STAT_ETHER_STATS_TX_NO_ERRORS\":\"0\",\"SAI_PORT_STAT_ETHER_TX_OVERSIZE_PKTS\":\"0\",\"SAI_PORT_STAT_IF_IN_BROADCAST_PKTS\":\"0\",\"SAI_PORT_STAT_IF_IN_DISCARDS\":\"0\",\"SAI_PORT_STAT_IF_IN_ERRORS\":\"0\",\"SAI_PORT_STAT_IF_IN_MULTICAST_PKTS\":\"0\",\"SAI_PORT_STAT_IF_IN_NON_UCAST_PKTS\":\"0\",\"SAI_PORT_STAT_IF_IN_OCTETS\":\"0\",\"SAI_PORT_STAT_IF_IN_UCAST_PKTS\":\"0\",\"SAI_PORT_STAT_IF_IN_UNKNOWN_PROTOS\":\"0\",\"SAI_PORT_STAT_IF_OUT_BROADCAST_PKTS\":\"0\",\"SAI_PORT_STAT_IF_OUT_DISCARDS\":\"0\",\"SAI_PORT_STAT_IF_OUT_ERRORS\":\"0\",\"SAI_PORT_STAT_IF_OUT_MULTICAST_PKTS\":\"0\",\"SAI_PORT_STAT_IF_OUT_NON_UCAST_PKTS\":\"0\",\"SAI_PORT_STAT_IF_OUT_OCTETS\":\"0\",\"SAI_PORT_STAT_IF_OUT_QLEN\":\"0\",\"SAI_PORT_STAT_IF_OUT_UCAST_PKTS\":\"0\",\"SAI_PORT_STAT_IP_IN_UCAST_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_0_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_0_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_1_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_1_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_2_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_2_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_3_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_3_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_4_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_4_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_5_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_5_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_6_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_6_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_7_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_7_TX_PKTS\":\"0\"}"
    >
  >
>

sync_response: true
```

If `COUNTERS_DB` is not supported shouldn't give it an error right away?

roweisch commented on sonic_data_client/db_client.go:
```diff
@@ -62,7 +62,7 @@ type Stream interface {
 
 // redis client connected to each DB
-var Target2RedisDb = make(map[string]*redis.Client)
+var Target2RedisDb = make(map[string]map[string]*redis.Client)
```
I am new to go, but could you explain why this map[string] is repeated here?

abdosi replied:
@roweisch yes. Basically we have {dbname} to {redis client}  mapping .. now we have {namespace, dbname} to {redis client} mapping.

abdosi commented on sonic_data_client/virtual_db.go:
```diff
@@ -122,145 +123,161 @@ func getPfcwdMap() (map[string]map[string]string, error) {
-		pfcwdName_map[name] = make(map[string]string)
-	}
+		for _, key := range resp {
+			name := key[13:]
```
this represent the name in key. Best place to see different key name used in in SONiC is to use to refer schema.
https://github.com/Azure/sonic-swss-common/blob/master/common/schema.h

abdosi commented on gnmi_server/server_test.go:
```diff
@@ -2212,206 +2306,202 @@ func TestCapabilities(t *testing.T) {
+	tlsConfig := &tls.Config{InsecureSkipVerify: true}
+	opts := []grpc.DialOption{grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))}
+
+	//targetAddr := "30.57.185.38:8080"
```
thanks, we can do cleanup as part of other PR.

abdosi commented:
> This is a comment for `db_client.go` line 132:
> 
> ```
> 	if prefix.GetTarget() == "COUNTERS_DB" {
> 		err = initCountersPortNameMap()
> 		if err != nil {
> 			return nil, err
> 		}
>                
>                // Code to initialize other maps...
> 	}
> ```
> 
> This code wouldn't run in case target has a asic name segment. I recall `COUNTERS_DB` is not supported for multi asic queries.
> However, `COUNTERS_DB/asic0` doesn't give me an error:
> 
> ```
> update: <
>   timestamp: 1623266624421030390
>   prefix: <
>     target: "COUNTERS_DB/asic0"
>   >
>   update: <
>     path: <
>       elem: <
>         name: "COUNTERS"
>       >
>       elem: <
>         name: "Ethernet0"
>       >
>     >
>     val: <
>       string_val: ""
>     >
>   >
> >
> 
> sync_response: true
> ```
> 
> Compare to the query without asic suffix (on the same environment):
> 
> ```
> update: <
>   timestamp: 1623266666110051137
>   prefix: <
>     target: "COUNTERS_DB"
>   >
>   update: <
>     path: <
>       elem: <
>         name: "COUNTERS"
>       >
>       elem: <
>         name: "Ethernet0"
>       >
>     >
>     val: <
>       json_ietf_val: "{\"SAI_PORT_STAT_ETHER_IN_PKTS_128_TO_255_OCTETS\":\"0\",\"SAI_PORT_STAT_ETHER_RX_OVERSIZE_PKTS\":\"0\",\"SAI_PORT_STAT_ETHER_STATS_TX_NO_ERRORS\":\"0\",\"SAI_PORT_STAT_ETHER_TX_OVERSIZE_PKTS\":\"0\",\"SAI_PORT_STAT_IF_IN_BROADCAST_PKTS\":\"0\",\"SAI_PORT_STAT_IF_IN_DISCARDS\":\"0\",\"SAI_PORT_STAT_IF_IN_ERRORS\":\"0\",\"SAI_PORT_STAT_IF_IN_MULTICAST_PKTS\":\"0\",\"SAI_PORT_STAT_IF_IN_NON_UCAST_PKTS\":\"0\",\"SAI_PORT_STAT_IF_IN_OCTETS\":\"0\",\"SAI_PORT_STAT_IF_IN_UCAST_PKTS\":\"0\",\"SAI_PORT_STAT_IF_IN_UNKNOWN_PROTOS\":\"0\",\"SAI_PORT_STAT_IF_OUT_BROADCAST_PKTS\":\"0\",\"SAI_PORT_STAT_IF_OUT_DISCARDS\":\"0\",\"SAI_PORT_STAT_IF_OUT_ERRORS\":\"0\",\"SAI_PORT_STAT_IF_OUT_MULTICAST_PKTS\":\"0\",\"SAI_PORT_STAT_IF_OUT_NON_UCAST_PKTS\":\"0\",\"SAI_PORT_STAT_IF_OUT_OCTETS\":\"0\",\"SAI_PORT_STAT_IF_OUT_QLEN\":\"0\",\"SAI_PORT_STAT_IF_OUT_UCAST_PKTS\":\"0\",\"SAI_PORT_STAT_IP_IN_UCAST_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_0_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_0_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_1_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_1_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_2_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_2_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_3_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_3_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_4_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_4_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_5_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_5_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_6_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_6_TX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_7_RX_PKTS\":\"0\",\"SAI_PORT_STAT_PFC_7_TX_PKTS\":\"0\"}"
>     >
>   >
> >
> 
> sync_response: true
> ```
> 
> If `COUNTERS_DB` is not supported shouldn't give it an error right away?
> 
> **UPDATE:**
> In fact it returns a proper error if there was a legit `COUNTERS_DB` query that proceeding the `COUNTERS_DB/asic0` query. That legit query must be initializing something that allows the v2r check for `COUNTERS_DB/asic0` return result (hence give the proper error message). Though, I didn't dig deeper what is that initialization that first query triggers.

Thanks @macikgozwa I have fixed this in latest commit. Also added unit test case for same.

abdosi commented on sonic_db_config/db_config.go:
```diff
@@ -1,135 +1,236 @@
+				panic(fmt.Errorf("Global Database config file is not valid(multiple include for same namespace!"))
+			}
+			//Ref:https://www.geeksforgeeks.org/filepath-join-function-in-golang-with-examples/
+			db_include_file := filepath.Join(dir, entry.(map[string]interface{})["include"].(string))
```
@macikgozwa the same file is used in other places also. So we should be fine.

abdosi commented:
/azp run

abdosi commented:
@lguohan @liushilongbuaa  can you plese help check azure pipeline build failure.

We are not able to down artifact for sonic-mgmt-common repo.

hui-ma commented:
See failure like below in utest. It is expected failure?
E0427 18:41:56.041761    9141 non_db_client.go:308] Failed to read '/etc/sonic/sonic_version.yml', Cannot access '/etc/sonic/sonic_version.yml'

macikgozwa commented:
> 
> 
> See failure like below in utest. It is expected failure?
> E0427 18:41:56.041761 9141 non_db_client.go:308] Failed to read '/etc/sonic/sonic_version.yml', Cannot access '/etc/sonic/sonic_version.yml'
> 
> @hui-ma this is negative test case I belive added by @macikgozwa

Yes, we have a negative case to test what happens when the version file is missing. @hui-ma @abdosi

abdosi commented:
/azp run
